bookkeeper-server/src/main/java/org/apache/bookkeeper/http/ListUnderReplicatedLedgerService.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.commons.lang.StringUtils;
+import org.apache.zookeeper.ZooKeeper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * HttpEndpointService that handle Bookkeeper Configuration related http request.
+ */
+public class ListUnderReplicatedLedgerService implements HttpEndpointService {
+
+    static final Logger LOG = LoggerFactory.getLogger(ListUnderReplicatedLedgerService.class);
+
+    protected ServerConfiguration conf;
+
+    public ListUnderReplicatedLedgerService(ServerConfiguration conf) {
+        Preconditions.checkNotNull(conf);
+        this.conf = conf;
+    }
+
+    /*
+     * Print the node which holds the auditor lock.
+     */
+    @Override
+    public HttpServiceResponse handle(HttpServiceRequest request) throws Exception {
+        HttpServiceResponse response = new HttpServiceResponse();
+        // parameter as this: ?missingreplica=<bookie_address>&excludingmissingreplica=<bookid_address>
+        Map<String, String> params = request.getParams();
+
+        if (HttpServer.Method.GET == request.getMethod()) {
+            final String includingBookieId;
+            final String excludingBookieId;
+            if (params != null && params.containsKey("missingreplica")) {
+                includingBookieId = params.get("missingreplica");
+            } else {
+                includingBookieId = null;
+            }
+            if (params != null && params.containsKey("excludingmissingreplica")) {
+                excludingBookieId = params.get("excludingmissingreplica");
+            } else {
+                excludingBookieId = null;
+            }
+            Predicate<List<String>> predicate = null;
+            if (!StringUtils.isBlank(includingBookieId) && !StringUtils.isBlank(excludingBookieId)) {
+                predicate = replicasList -> (replicasList.contains(includingBookieId)
+                  && !replicasList.contains(excludingBookieId));
+            } else if (!StringUtils.isBlank(includingBookieId)) {
+                predicate = replicasList -> replicasList.contains(includingBookieId);
+            } else if (!StringUtils.isBlank(excludingBookieId)) {
+                predicate = replicasList -> !replicasList.contains(excludingBookieId);
+            }
+
+            ZooKeeper zk = ZooKeeperClient.newBuilder()
+                .connectString(conf.getZkServers())
+                .sessionTimeoutMs(conf.getZkTimeout())
+                .build();
+
+            try {
+                List<Long> outputLedgers = Lists.newArrayList();
+                LedgerManagerFactory mFactory = LedgerManagerFactory.newLedgerManagerFactory(conf, zk);
+                LedgerUnderreplicationManager underreplicationManager = mFactory.newLedgerUnderreplicationManager();
+                Iterator<Long> iter = underreplicationManager.listLedgersToRereplicate(predicate);
+
+                while (iter.hasNext()) {
+                    outputLedgers.add(iter.next());
```

Comment:
If we have a large number of underreplicated ledgers, this will take a long time to run. What we do in Twitter is to run the underreplicated ledger analyzer periodically in auditor and store a list of underreplicated ledgers, so Http Endpoint will just return the list directly.

Reply:
Thanks. will consider and change it.